python/ray/serve/_private/controller.py

```diff
@@ -375,6 +384,109 @@ def _update_proxy_nodes(self):
         new_proxy_nodes.add(self._controller_node_id)
         self._proxy_nodes = new_proxy_nodes
 
+    def _emit_deployment_autoscaling_snapshots(self) -> None:
+        """Emit a structured snapshot log per autoscaling-enabled deployment."""
+        try:
+            app_statuses = self.application_state_manager.list_app_statuses()
+            for app_name in app_statuses.keys():
+                deployment_details = (
+                    self.application_state_manager.list_deployment_details(app_name)
+                )
+                for dep_name, details in deployment_details.items():
+                    autoscaling_config = details.deployment_config.autoscaling_config
+                    if not autoscaling_config:
+                        continue
+                    current_replicas = len(getattr(details, "replicas", []) or [])
+                    raw_target = getattr(details, "target_num_replicas", None)
+                    current_target = (
+                        int(raw_target) if raw_target is not None else current_replicas
+                    )
+                    dep_id = DeploymentID(name=dep_name, app_name=app_name)
+                    snapshot = (
+                        self.autoscaling_state_manager.get_observability_snapshot(
+                            dep_id, current_target
+                        )
+                    )
+                    current_replicas = snapshot["current_replicas"]
+                    proposed_replicas = snapshot["proposed_replicas"]
+                    min_repl_adj = snapshot["min_replicas"]
+                    max_repl_adj = snapshot["max_replicas"]
+                    total_requests = snapshot["total_requests"]
+
+                    if isinstance(autoscaling_config, dict):
+                        policy_name = autoscaling_config.get("policy", "default")
+                        look_back_period_s = autoscaling_config.get(
+                            "look_back_period_s"
+                        )
+                    else:
+                        policy_name = getattr(autoscaling_config, "name", "default")
+                        look_back_period_s = getattr(
+                            autoscaling_config, "look_back_period_s", None
+                        )
```

Comment:
this seems odd, why can't we assume `autoscaling_config` as either a dict or pydantic object? why do we need this dual logic for accessing attributes of autoscaling_config